lib/SILOptimizer/Differentiation/JVPCloner.cpp

```diff
@@ -359,6 +368,96 @@ class JVPCloner::Implementation final
         type.getCategory());
   }
 
+  void addPayloadPointerArgument(SILBasicBlock *bb, SILBasicBlock *newBb) {
+    SILType builtinRawPtrTy = SILType::getRawPointerType(getASTContext());
+    newBb->createPhiArgument(builtinRawPtrTy, ValueOwnershipKind::Owned);
+  }
+
+  SILBasicBlock *createTrampolineBlock(SILBasicBlock *bb) {
+    auto *trampBlock = jvp->createBasicBlock();
+    for (auto &arg : bb->getArguments())
+      trampBlock->createPhiArgument(arg->getType(), arg->getOwnershipKind());
+    addPayloadPointerArgument(bb, trampBlock);
+    return trampBlock;
+  }
+
+  void addBranchInstToTrampolineBlock(SILLocation loc, TermInst *origTermInst,
+                                      SILBasicBlock *origSuccBB) {
+    auto *bb = origTermInst->getParent();
+    auto *trampBB = jvpTrampolineBBMap.lookup({bb, origSuccBB});
+    assert(trampBB && "Trampoline block should exist");
+
+    auto *jvpTermInst = getOpBasicBlock(bb)->getTerminator();
+    SILBuilderWithScope trampBuilder(trampBB, jvpTermInst);
+    getBuilder().setCurrentDebugScope(trampBuilder.getCurrentDebugScope());
+
+    // Step 1: Create enum with the undef payload for the next successor bb.
+    auto *succStruct = differentialInfo.getLinearMapStruct(origSuccBB);
+    auto succStructLoweredTy =
+        getLoweredType(succStruct->getDeclaredInterfaceType());
+    auto *undefStruct = SILUndef::get(succStructLoweredTy, getModule(),
+                                      ValueOwnershipKind::Owned);
+    auto *curEnum = differentialInfo.getBranchingTraceDecl(bb);
+    auto enumLoweredTy = getNominalDeclLoweredType(curEnum);
+    auto *enumEltDecl =
+        differentialInfo.lookUpBranchingTraceEnumElement(origSuccBB, bb);
+    assert(enumEltDecl && "Successor enum element declaration not found");
+
+    auto *succEnum =
+        trampBuilder.createEnum(loc, undefStruct, enumEltDecl, enumLoweredTy);
```

Comment:
Based on https://github.com/apple/swift/blob/master/docs/SIL.rst#enums, and in particular the "init_enum_data_addr" and "inject_enum_addr" sections, I think that this strategy of initializing the enum with an undefined payload and then writing to the "init_enum_data_addr" address is undefined behavior.

Given what's written in that documentation, I don't see any way to avoid walking backwards through the trace to finish up the initializations.

This will be a pretty big change which you probably won't have time to do in your remaining time with us.

Reply:
I wrote up a detailed explanation and question here: https://forums.swift.org/t/possible-to-inject-enum-case-before-initializing-payload-in-sil/40764

Reply:
Wow, interesting. Thanks!